pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/ReplicaGroupRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,491 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ReplicaGroupRebalanceSegmentStrategy.class);
+  private static final boolean DEFAULT_REPLICA_GROUP_BOOTSTRAP = false;
+
+  public ReplicaGroupRebalanceSegmentStrategy(HelixManager helixManager) {
+    super(helixManager);
+  }
+
+  @Override
+  public PartitionAssignment rebalancePartitionAssignment(IdealState idealState, TableConfig tableConfig,
+      Configuration rebalanceUserConfig) {
+    // Currently, only offline table is supported
+    if (tableConfig.getTableType() == CommonConstants.Helix.TableType.REALTIME) {
+      throw new UnsupportedOperationException("Replica group rebalancer is currently not supported for Realtime table");
+    }
+
+    String tableNameWithType =
+        TableNameBuilder.forType(tableConfig.getTableType()).tableNameWithType(tableConfig.getTableName());
+    LOGGER.info("Rebalancing replica group partition assignment for table {}", tableNameWithType);
+
+    ReplicaGroupPartitionAssignmentGenerator partitionAssignmentGenerator =
+        new ReplicaGroupPartitionAssignmentGenerator(_propertyStore);
+    boolean bootstrap = rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.REPLICA_GROUP_BOOTSTRAP,
+        DEFAULT_REPLICA_GROUP_BOOTSTRAP);
+
+    // Compute new replica group partition assignment
+    ReplicaGroupPartitionAssignment newPartitionAssignment =
+        computeNewReplicaGroupMapping(tableConfig, partitionAssignmentGenerator, bootstrap);
+
+    boolean dryRun = rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.DRYRUN, DEFAULT_DRY_RUN);
+    if (!dryRun) {
+      LOGGER.info("Updating replica group partition assignment for table {}", tableNameWithType);
+      partitionAssignmentGenerator.writeReplicaGroupPartitionAssignment(newPartitionAssignment);
+    } else {
+      LOGGER.info("Dry run. Skip writing replica group partition assignment to property store");
+    }
+
+    return newPartitionAssignment;
+  }
+
+  @Override
+  public IdealState rebalanceIdealState(IdealState idealState, TableConfig tableConfig,
+      Configuration rebalanceUserConfig, PartitionAssignment newPartitionAssignment) {
+
+    ReplicaGroupPartitionAssignment newReplicaGroupPartitionAssignment =
+        (ReplicaGroupPartitionAssignment) newPartitionAssignment;
+    String tableNameWithType =
+        TableNameBuilder.forType(tableConfig.getTableType()).tableNameWithType(tableConfig.getTableName());
+    boolean bootstrap = rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.REPLICA_GROUP_BOOTSTRAP,
+        DEFAULT_REPLICA_GROUP_BOOTSTRAP);
+
+    // Rebalancing segments
+    IdealState newIdealState =
+        rebalanceSegments(idealState, tableConfig, newReplicaGroupPartitionAssignment, bootstrap);
+    int numReplication = tableConfig.getValidationConfig().getReplicationNumber();
+
+    // update if not dryRun
+    boolean dryRun = rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.DRYRUN, DEFAULT_DRY_RUN);
```

Comment:
Will you be making the change similar to DefaultRebalanceSegmentsStrategy to move rebalanceSegments inside the retry?

Reply:
Yes, I'm following DefaultRebalanceSegmentsStrategy